server/src/main/java/org/cloudfoundry/identity/uaa/scim/vmidentity/VmidentityScimGroupMembershipManager.java

```diff
@@ -0,0 +1,343 @@
+            String systemDomain = VmidentityUtils.getSystemDomain(tenant, this.idmClient);
+            Group group = this.idmClient.findGroup(tenant, groupId);
+            if (systemDomain.equalsIgnoreCase(group.getId().getDomain()) == false) {
+                logger.error("Cannot modify group membership for groups in non-system domain. " + groupId);
+                throw new InvalidScimResourceException(
+                        "Cannot modify group membership for groups in non-system domain.");
+            }
+
+            Principal idmPrincipal = null;
+            if (member.getType() == ScimGroupMember.Type.USER) {
+                idmPrincipal = this.idmClient.findUser(tenant, member.getMemberId());
+                if (idmPrincipal != null) {
+                    this.idmClient.addUserToGroup(tenant, idmPrincipal.getId(), group.getName());
+                } else {
+                    logger.error("user not found  " + member.getMemberId());
+                    throw new ScimResourceNotFoundException("User not found: " + member.getMemberId());
+                }
+            } else if (member.getType() == ScimGroupMember.Type.GROUP) {
+                idmPrincipal = this.idmClient.findGroup(tenant, member.getMemberId());
+                if (idmPrincipal != null) {
+                    this.idmClient.addGroupToGroup(tenant, idmPrincipal.getId(), group.getName());
+                } else {
+                    logger.error("group not found  " + member.getMemberId());
+                    throw new ScimResourceNotFoundException("Group not found: " + member.getMemberId());
+                }
+            } else {
+                throw new InvalidArgumentException(String.format("Invalid ScimMemberType '%s'", member.getType()));
+            }
+
+            return member;
+        } catch (com.vmware.identity.idm.MemberAlreadyExistException ex) {
+            logger.debug(String.format("Scim group member already exists '%s'.", member.getMemberId()));
+            throw new MemberAlreadyExistsException(
+                    String.format("Scim group member already exists '%s'.", member.getMemberId()));
+        } catch (InvalidPrincipalException ex) {
+            logger.debug(String.format("Scim resource not found '%s'.", ex.getPrincipal()));
+            throw new ScimResourceNotFoundException(String.format("Scim resource not found '%s'.", ex.getPrincipal()));
+        } catch (Exception ex) {
+            logger.debug(String.format("addMember '%s', '%s' failed", groupId, member.getMemberId()), ex);
+            throw new IllegalStateException(String.format("addMember '%s', '%s' ", groupId, member.getMemberId()), ex);
+        }
+    }
+
+    @Override
+    public List<ScimGroupMember> getMembers(String groupId, String filter, boolean includeEntities)
+            throws ScimResourceNotFoundException {
+        // TODO: Figure out how to filter these...
```

Comment:
So it turned out that simply being able to filter on the 'member' attribute in LDAP wasn't for this method in particular. Rather, we want to apply the filter to the group's members, which in turn required a pretty expensive operation: fetching a group, fetching all of the users/groups that conform to the filter, then finding the intersect between the group's membership and everything we retrieved.

Ideally we'd be able to manually maintain a 'memberOf' attribute and then we could easily filter on that (e.g. append a `(memberOf=<groupId>)` clause to the query). Currently the attribute is computed and thus ineligible for querying against, but internally we could maintain that, I think.

In any case, this call is explicitly only called with the `filter` parameter as `null` (internally to UAA), so I wasn't too worried about implementing such a cumbersome solution. We can always implement it later if needed - perhaps I can add some additional exception handling if `filter` isn't `null` or the empty string, so that way it'll get caught.